helix-core/src/main/java/org/apache/helix/messaging/handling/AsyncCallbackService.java

```diff
@@ -113,12 +104,19 @@ public HelixTaskResult handleMessage() throws InterruptedException {
           + _correlationId);
 
       AsyncCallback callback = _callbackMap.get(_correlationId);
-      synchronized (callback) {
-        callback.onReply(_message);
-        if (callback.isDone()) {
-          _logger.info("Removing finished callback, correlationid:" + _correlationId);
-          _callbackMap.remove(_correlationId);
+      if (callback != null) {
+        synchronized (callback) {
+          callback.onReply(_message);
+          if (callback.isDone()) {
+            _logger.info("Removing finished callback, correlationid:" + _correlationId);
+            _callbackMap.remove(_correlationId);
+          }
         }
+      } else {
+        String msg = "Message " + _message.getMsgId()
+            + " does not have correponding callback. Probably timed out already. Correlation id: "
+            + _correlationId;
+        _logger.warn(msg);
```

Comment:
A question regarding the description, if the error is just logged here, meaning the users won't really know there is a mismatch. If the callback is not done, they won't be able to distinguish between this error and other errors. Is this the expected behavior?

Reply:
You mean they wouldn't be able to distinguish "stale messages" vs "messages that don't have callbacks set up correctly"? I think these cases are logically the same and can't be easily distinguished code-level. If we are loosening up on one case, it's going to affect the other. 

On the other hand, callback registration is done in our code and naturally there should be a callback before sending messages that require replies. I imagine the previous check was there for cautionary purpose, not that messages are commonly created with incorrect correlation ids.

Reply:
Actually I meant this missing ID exception and exceptions in the later functions in onReply. Do we need to consider distinguishing between them? like in your test `Assert.assertFalse(callback.isDone())`. Any other errors may happen in middle?

Reply:
I'm not sure I follow: this is not a "try" block, this is an "if" block. There's no exception handling here.

Reply:
All i would like to ask is whether removing the previous thrown exception will cause any information lost. If it's ok, then I'm fine.

Reply:
I see what you mean now: you're saying in both cases, callback.isDone() == false, and it can't be distinguished whether the callback failed or there's no callback registered. 

If the callback fails due to some exception like you said, the task result will reflect that (setException(e), setInterrupted(true)). It's more difficult to tell the "callback not registered" case, but since customers don't want exceptions raised, there isn't any better way here.

Reply:
Yep, that's what I meant. I think it's not a big concern.